crypto/zkp/schemes/cl/user.go

```diff
@@ -26,22 +26,45 @@ import (
 )
 
 type User struct {
-	ParamSizes     *CLParamSizes
-	PubKey         *PubKey
-	PedersenParams *commitments.PedersenParams               // for pseudonyms - nym is a commitment to the master secret
-	Committers     map[string]*commitments.PedersenCommitter // for generating nyms
-	masterSecret   *big.Int
-	attrs          []*big.Int
+	ParamSizes         *CLParamSizes
+	PubKey             *PubKey
+	PedersenParams     *commitments.PedersenParams               // for pseudonyms - nym is a commitment to the master secret
+	Committers         map[string]*commitments.PedersenCommitter // for generating nyms
+	masterSecret       *big.Int
+	knownAttrs         []*big.Int                              // attributes that are known to the credential receiver and issuer
+	committedAttrs     []*big.Int                              // attributes for which the issuer knows only commitment
+	hiddenAttrs        []*big.Int                              // attributes which are known only to the credential receiver
+	attrsCommitters    []*commitments.DamgardFujisakiCommitter // committers for committedAttrs
+	commitmentsOfAttrs []*big.Int                              // commitments of committedAttrs
 }
 
-func NewUser(clParamSizes *CLParamSizes, clPubKey *PubKey, pedersenParams *commitments.PedersenParams) *User {
-	return &User{
-		ParamSizes:     clParamSizes,
-		PubKey:         clPubKey,
-		Committers:     make(map[string]*commitments.PedersenCommitter),
-		PedersenParams: pedersenParams,
-		attrs:          []*big.Int{big.NewInt(7), big.NewInt(6), big.NewInt(5)}, // TODO attributes should be read from somewhere and the structure should be loaded too
+func NewUser(clParamSizes *CLParamSizes, clPubKey *PubKey, pedersenParams *commitments.PedersenParams,
+	knownAttrs, committedAttrs, hiddenAttrs []*big.Int) (*User, error) {
+
+	attrsCommitters := make([]*commitments.DamgardFujisakiCommitter, len(committedAttrs))
+	commitmentsOfAttrs := make([]*big.Int, len(committedAttrs))
+	for i, attr := range committedAttrs {
+		committer := commitments.NewDamgardFujisakiCommitter(clPubKey.N1, clPubKey.H, clPubKey.G,
+			clPubKey.N1, clParamSizes.SecParam)
+		com, err := committer.GetCommitMsg(attr)
+		if err != nil {
+			return nil, fmt.Errorf("error when creating Pedersen commitment: %s", err)
+		}
+		commitmentsOfAttrs[i] = com
+		attrsCommitters[i] = committer
 	}
+
+	return &User{
+		ParamSizes:         clParamSizes,
+		PubKey:             clPubKey,
+		Committers:         make(map[string]*commitments.PedersenCommitter),
+		PedersenParams:     pedersenParams,
+		knownAttrs:         knownAttrs,
+		committedAttrs:     committedAttrs,
+		hiddenAttrs:        hiddenAttrs,
+		commitmentsOfAttrs: commitmentsOfAttrs,
+		attrsCommitters:    attrsCommitters,
+	}, nil
 }
 
```

Comment:
Since `GenerateMasterSecret` doesn't return any value, but merely sets the `masterSecret` field, I propose that you:
1. Copy L71 to and move initialization of the `masterSecret` field to `NewUser` function. 
1. Remove the `GenerateMasterSecret` method.

Why? Well, what would happen if I created a `User` but forgot to explicitly call its `GenerateMasterSecret` method? The `masterSecret` field would be `nil`, and thus the call to `GenerateNym` would panic. This could never happen with the proposed change.